raco/backends/myria/myria.py

```diff
@@ -1721,6 +1824,193 @@ def fire(self, op):
+            if isinstance(upper_op, algebra.UnaryOperator):
+                upper_op.input = new_consumer
+            elif isinstance(upper_op, algebra.BinaryOperator):
+                if upper_op.left == parent_map[id(producer)][0]:
+                    upper_op.left = new_consumer
+                else:
+                    upper_op.right = new_consumer
+            else:
+                for idx, t in enumerate(upper_op.args):
+                    if upper_op.args[idx] == parent_map[id(producer)][0]:
+                        upper_op.args[idx] = new_consumer
+                        break
+            op.args[idx] = producer
+        return op
+
+
+class DoUntilConvergence(rules.Rule):
+
+    @staticmethod
+    def replace_scan_with_idb_consumer(expr, idb_controllers, idb_producers):
+        if isinstance(expr, algebra.ZeroaryOperator):
+            return
+        for idx, op in enumerate(expr.children()):
+            if isinstance(op, algebra.ScanIDB):
+                if op.name not in idb_producers:
+                    idb_producers[op.name] = \
+                        MyriaSplitProducer(idb_controllers[op.name])
+                consumer = MyriaConsumer(None, idb_producers[op.name])
+                if isinstance(expr, algebra.UnaryOperator):
+                    expr.input = consumer
+                elif isinstance(expr, algebra.BinaryOperator):
+                    if idx == 0:
+                        expr.left = consumer
+                    else:
+                        expr.right = consumer
+                elif isinstance(expr, algebra.NaryOperator):
+                    expr.args[idx] = consumer
+                if isinstance(expr, algebra.Select) and \
+                        isinstance(expr.condition, expression.boolean.GTEQ):
+                    (group_list, agg) = \
+                        idb_controllers[op.name].get_group_agg()
+                    if isinstance(agg, expression.aggregate.COUNTALL):
+                        agg.threshold = expr.condition.right.value
+                        expr.condition = None
+            else:
+                DoUntilConvergence.replace_scan_with_idb_consumer(
+                    op, idb_controllers, idb_producers)
+
+    def fire(self, op):
+        if not isinstance(op, algebra.UntilConvergence):
+            return op
+        if hasattr(op, "has_been_transformed"):
+            if op.has_been_transformed:
+                return op
+
+        eos_controller = MyriaEOSController()
+        eos_consumers = []
+        idb_controllers = {}
+        for idb_controller in op.args:
+            idb_controller.children()[2] = MyriaConsumer(None, eos_controller)
+            eos_consumers.append(MyriaConsumer(None, idb_controller))
+            idb_controllers[idb_controller.name] = idb_controller
+        if len(eos_consumers) == 1:
+            eos_controller.input = eos_consumers[0]
+        else:
+            eos_controller.input = MyriaUnionAll(eos_consumers)
```

Comment:
I don't expect `UnionAll` to have only one child in all the other places since it's constructed from `[...] + [...]`, so adding another rule seems unnecessary.